### PR TITLE
GetUpcomingInvoice for BillingService (2)

### LINKED ISSUE #11692: Implement BillingService.GetLatestInvoice
Currently, when showing spending notifications to users we use the `UsageService.ListUsage` RPC and we sum up all the records. This is suboptimal, but more importantly it is the wrong data. Instead, we need to be looking at the current value of an Invoice (unpaid).

To that end, we want to add a new RPC to the `BillingService`. The RPC should be able to do the following:
* Given a UserID, or TeamID, it find details about the latest invoice and retrieves the current value of that invoce (along with more info if needed)

### PR DESCRIPTION
**Pull request by AlexTugarev**
## Description
This PR implements ~`GetLatestInvoice`~  `GetUpcomingInvoice`. The rename is done intentionally, to a) align with terminology used in stripe, and b) to distinguish between the next invoice and an already paid invoice. The upcoming invoice contains the the amount of credits currently used (🤞🏻 after discounts), which can be compared with the spending limit configured.

- [x] rename GetLatestInvoice to GetUpcomingInvoice

## Related Issue(s)

Fixes #11692 

## How to test
Setup UBB in for a team and lower the spending limit to 1 or 2 credits for a simple test, then start a workspace or two to hit that limit. After a while, cannot tell it precisely, you'll need to reload the dashboard to refresh the spending limit checks. In the `server` component you'll find a log message on spending limit events, and the known dashboard notification should be visible:
Screen Shot 2022-08-23 at 16 21 48

## Release Notes

```release-note
NONE
```

## Werft options:

- [x] /werft with-preview
- [x] /werft with-payment

### CODE REVIEW

File: components/server/ee/src/billing/billing-service.ts [hunk 1]
```diff
@@ -40,50 +46,53 @@ export class BillingService {
...
-        if (totalUsage >= costCenter.spendingLimit) {
+        if (!(await this.isSpendingLimitCheckEnabled(user))) {
+            return {
+                reached: false,
+                attributionId,
+            };
+        }
+
+        const upcomingInvoice = await this.getUpcomingInvoice(attributionId);
+        const currentUsage = upcomingInvoice.getCredits();
```
Comment (easyCZ) [inline components/server/ee/src/billing/billing-service.ts:57]:
```suggestion
        const currentInvoiceCredits = upcomingInvoice.getCredits();
```

File: components/usage/pkg/stripe/stripe.go [hunk 2]
```diff
@@ -122,6 +137,47 @@ func (c *Client) updateUsageForCustomer(ctx context.Context, customer *stripe.Cu
 	}, nil
 }
 
+// GetUpcomingInvoice fetches the upcoming invoice for the given team or user id.
+func (c *Client) GetUpcomingInvoice(ctx context.Context, in *v1.GetUpcomingInvoiceRequest) (*StripeInvoice, error) {
```
Comment (easyCZ) [inline components/usage/pkg/stripe/stripe.go:141]: The stripe client in this package shouldn't be taking on a dependency from the API. It's tightly coupling the Stripe client to the API when they should be each able to evolve independently. 

In this case, I'd supply either teamID or userID as a string to break this coupling.

File: components/usage/pkg/stripe/stripe.go [hunk 3]
```diff
@@ -122,6 +137,47 @@ func (c *Client) updateUsageForCustomer(ctx context.Context, customer *stripe.Cu
...
+			Query:   query,
+			Expand:  []*string{stripe.String("data.subscriptions")},
+			Context: ctx,
+		},
+	}
+	iter := c.sc.Customers.Search(searchParams)
+	if iter.Err() != nil && !iter.Next() {
+		return nil, ErrorCustomerNotFound
+	}
+	customer := iter.Customer()
```
Comment (easyCZ) [inline components/usage/pkg/stripe/stripe.go:160]: What if the iterator contains multiple results?

**Comment by easyCZ** [comment]
What's the plan for dealing with the scenario when the `GetUpcomingInvoice` RPC is not implemented on the usage component when server rolls out?

File: components/usage/pkg/stripe/stripe.go (hunk 3 above)
Comment (AlexTugarev) [reply components/usage/pkg/stripe/stripe.go:160]: raising then

**Review by easyCZ — APPROVED** [review]

**Pull request merged**